### PR TITLE
Add Raw18

### LINKED ISSUE #6210: raw18
### Source name

raw18.net

### Source link

https://raw18.net/

### Source language

japanese

### Other details

18+NSFW = yes

### Acknowledgements

- [X] I have checked that the extension does not already exist by searching the [extension listing](https://keiyoushi.github.io/extensions/) and verified it does not appear there.
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a meaningful title with the source name.
- [X] I will fill out all of the requested information in this form.

### 

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/keiyoushi/extensions-source/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by TeflonOfJoy**
Closes #6210

Checklist:

- [ ] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [ ] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

File: src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    override val genresUrlDelimiter = "="
+    override fun popularMangaSelector() = "div.items article.item"
+    override fun popularMangaNextPageSelector() = "li:nth-last-child(2) a.page-link"
+    override fun searchMangaSelector() = popularMangaSelector()
+    override fun mangaDetailsParse(document: Document): SManga {
+        return SManga.create().apply {
+            document.select("article#item-detail").let { info ->
+                author = info.select("li.author p.col-xs-8").text()
+                status = when {
+                    info.select("li.status p.col-xs-8").text().contains("Ongoing", true) -> SManga.ONGOING
```
Comment (vetleledaal) [inline src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt]: Assuming you expect more status values, I would move parts of it into a variable.

```kt
val x = info.select("li.status p.col-xs-8").text()
```

File: src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.Locale
+
+class Raw18 : WPComics("Raw18", "https://raw18.net", "ja", SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'", Locale.JAPANESE), null) {
+    override val searchPath = "search/manga"
+    override val genresUrlDelimiter = "="
+    override fun popularMangaSelector() = "div.items article.item"
+    override fun popularMangaNextPageSelector() = "li:nth-last-child(2) a.page-link"
+    override fun searchMangaSelector() = popularMangaSelector()
+    override fun mangaDetailsParse(document: Document): SManga {
+        return SManga.create().apply {
```
Comment (vetleledaal) [inline src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt:30]: For the whole function, use `.selectFirst()` if you only expect one matching element.

File: src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                is StatusFilter -> filter.toUriPart()?.let { url.addQueryParameter("status", it) }
+                else -> {}
+            }
+        }
+        url.apply {
+            addQueryParameter(queryParam, query)
+            addQueryParameter("page", page.toString())
+        }
+        return GET(url.build(), headers)
+    }
```
Comment (vetleledaal) [inline src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt:55]: Consider something like:
```kt
override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request {
    val url = "$baseUrl/search/manga".toHttpUrl().newBuilder().apply {
        filters.ifEmpty { getFilterList() }.forEach { filter ->
            when (filter) {
                is GenreFilter -> filter.toUriPart()?.let { addQueryParameter("genre", it) }
                is StatusFilter -> filter.toUriPart()?.let { addQueryParameter("status", it) }
                else -> {}
            }
        }
        addQueryParameter(queryParam, query)
        addQueryParameter("page", page.toString())
    }.build()

    return GET(url, headers)
}
```

File: src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        val minuteWords = listOf("minute", "分")
+        val hourWords = listOf("hour", "時間")
+        val dayWords = listOf("day", "日")
+        val weekWords = listOf("week", "週間")
+        val monthWords = listOf("month", "月")
+        val chapterDate = element.select("div.col-xs-4").text()
+        return SChapter.create().apply {
+            element.select("a").let {
+                name = it.text()
+                setUrlWithoutDomain(it.attr("href"))
```
Comment (vetleledaal) [inline src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt]:
```suggestion
                setUrlWithoutDomain(it.absUrl("href"))
```

`it` must be of type `Element`, not `Elements`. You can use `document.selectFirst()` for that

File: src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            try {
+                val regex = Regex("([0-9]+)(.+)")
```
Comment (vetleledaal) [inline src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt]: Extract regex so it doesn't have to be recompiled each time. either as private variable or in the companion object.

File: src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        val dayWords = listOf("day", "日")
+        val weekWords = listOf("week", "週間")
+        val monthWords = listOf("month", "月")
+        val chapterDate = element.select("div.col-xs-4").text()
+        return SChapter.create().apply {
+            element.select("a").let {
+                name = it.text()
+                setUrlWithoutDomain(it.attr("href"))
+            }
+            try {
```
Comment (vetleledaal) [inline src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt]: Extract whole relative date parsing functionality into a private function. A bit better separation of concerns.

File: src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            }
+        }
+    }
+    override fun getStatusList(): List<Pair<String?, String>> {
+        return listOf(
+            Pair(null, "全て"),
+            Pair("ongoing", "Ongoing"),
+        )
+    }
+    override fun parseGenres(document: Document): List<Pair<String?, String>> {
```
Comment (vetleledaal) [inline src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt:124]: Why not wrap whole thing in `buildList {}`? Something like:

```kt
add(null to "全てのジャンル")
document.select(genresSelector).mapTo(this) { element ->
    element.attr("href")
        .removeSuffix("/")
        .substringAfterLast(genresUrlDelimiter) to element.text()
}
```

**Review by vetleledaal — CHANGES_REQUESTED** [review]

File: src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt (hunk 1 above)
Comment (TeflonOfJoy) [reply src/ja/raw18/src/eu/kanade/tachiyomi/extension/ja/raw18/Raw18.kt]: opted for toStatus() from wpcomics

**Review by vetleledaal — APPROVED** [review]

**Pull request merged**